### DISCUSSION TITLE
login_required! not working fully with nested routes

### DISCUSSION THREAD

**Discussion by jonner** [maxcountryman/axum-login#152] [Q&A]
If I have a route that's inside a nested router, and I use `login_required!` to try to protect it, it will pass along the nested uri as the `next` parameter rather than the full url.  For example, with the following untested code snippet:

```rust
let nested_router = Router::new().route("/new", get(new_thing))
    .route_layer(login_required!(Backend, login_url = "/login"));

let main_router = Router::new()
    .nest("/foo", nested_router);
```

When a person is not logged in and visits `/foo/new`, it sends them to `/login?next=%2fnew`, rather than `/login?next=%2ffoo%2fnew`. In other words, it is trying to redirect them back to `/new` after login rather than `/foo/new`. So the nested part of the URL gets lost. Is this supposed to work?

**Comment by maxcountryman** [comment]
The value of next is based on the request URI, as we can see [here](https://github.com/maxcountryman/axum-login/blob/009c5a29083f78928e79063e8dea13c6e6581de2/axum-login/src/middleware.rs#L139).

I'm surprised that nesting would affect this and it seems odd that the URI is truncated in that way. I wonder if this is correct behavior for axum.

Regardless, this isn't doing what we would want.

**Reply by maxcountryman** [reply]
The axum docs mention this: https://docs.rs/axum/latest/axum/routing/struct.Router.html#how-the-uri-changes

**Reply by jonner** [reply]
So, would you like me to open a bug report if this is not doing what you expect?

**Reply by maxcountryman** [reply]
I believe this should address the issue: https://github.com/maxcountryman/axum-login/pull/153

In a separate branch I'm working on a test suite for the middleware, so I'll cut a patch release and then follow up with a fix for a separate issue as well as a test case covering nested routes.

**Reply by jonner** [reply]
Thanks. That was quick.

**Reply by maxcountryman** [reply]
I've published v0.11.1. Do let us know if that doesn't address the issue.

**Reply by jonner** [reply]
Seems to work, thank you!

**Reply by maxcountryman** [reply]
I've also added [a test case](https://github.com/maxcountryman/axum-login/blob/e57e778170fef5ac560af2340706f51d0c2a8043/axum-login/src/middleware.rs#L728-L747) for this now.

Please let us know if you run into any other issues and we'll be happy to take a look.
